buster/formatters/documents.py

```diff
@@ -50,12 +89,59 @@ def format(
             )
             matched_documents = matched_documents.iloc[:num_preserved_docs]
 
+        documents_str = f"<DOCUMENTS>{documents_str}<\\DOCUMENTS>"
+
         return documents_str, matched_documents
 
 
-def documents_formatter_factory(tokenizer: Tokenizer, max_tokens: int, formatter: str) -> DocumentsFormatter:
-    return DocumentsFormatter(
-        tokenizer=tokenizer,
-        max_tokens=max_tokens,
-        formatter=formatter,
-    )
+@dataclass
+class DocumentsFormatterJSON(DocumentsFormatter):
+    """
+    Formatter class to convert matched documents into a JSON format.
+
+    Attributes:
+    - tokenizer (Tokenizer): Tokenizer instance to count tokens in the documents.
+    - max_tokens (int): Maximum allowed tokens for the formatted documents.
+    - columns (list[str]): List of columns to include in the JSON format.
+    """
+
+    tokenizer: Tokenizer
+    max_tokens: int
+    columns: list[str]
+
+    def format(self, matched_documents: pd.DataFrame) -> tuple[str, pd.DataFrame]:
+        """
+        Format the matched documents into a JSON format.
+
+        If the total tokens exceed max_tokens, documents are omitted one at a time until it fits the limit.
+
+        Parameters:
+        - matched_documents (pd.DataFrame): DataFrame containing the matched documents to be formatted.
+
+        Returns:
+        - tuple[str, pd.DataFrame]: A tuple containing the formatted documents as a JSON string and
+                                    the possibly truncated matched documents DataFrame.
+        """
+
+        max_tokens = self.max_tokens
+        documents_str = matched_documents[self.columns].to_json(orient="records")
+        token_count, _ = self.tokenizer.num_tokens(documents_str, return_encoded=True)
+
+        while token_count > max_tokens:
+            # Truncated too much, no documents left, raise an error
+            if len(matched_documents) == 0:
+                raise ValueError(
+                    f"Could not truncate documents to fit {max_tokens=}. Consider increasing max_tokens or decreasing chunk lengths."
+                )
+
+            # Too many tokens, drop a document and try again.
+            matched_documents = matched_documents.iloc[:-1]
+            documents_str = matched_documents[self.columns].to_json(orient="records")
+            token_count, _ = self.tokenizer.num_tokens(documents_str, return_encoded=True)
```

Comment:
Can we factor out the truncation code to the parent class? It seems the same for the HTML one

Reply:
its actually not the same ... for the HTML one, we can simply "cut" a source, in this one it's a bit trickier, but maybe we just force them to be the same for both?

Reply:
Let's leave them separate for now, but if we ever add a third formatter it might be worth forcing them to be identical

Reply:
ill add a TODO for now